### ISSUE TITLE
Return type cannot depend on a component of an argument

### ISSUE THREAD

**Issue by duvenaud** [google-research/dex-lang#286]
I decided to try Dex on graphs:
```
data Graph a:Type =
  MkGraph n:Type nodes:(n=>a) m:Type edges:(m=>(n & n))
```

However, when I try to write a function using this typeclass, it looks like type unification is failing:
```
def graphToAdjacencyMatrix ((MkGraph n nodes m edges):Graph a) : n=>n=>Bool =
  init = for i j. False
  snd $ withState init \mRef.
    for i:m.
      (from, to) = edges.i
      mRef!from!to := True
```
gives
```
Type error:
Expected: (n => n => Bool)
  Actual: (n1 => n1 => Bool)

  snd $ withState init \mRef.
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

However if I bypass the typeclass and simply provide a table of edges, it typechecks fine:
```
def edgesToAdjacencyMatrix (edges:(m=>(n & n))) : n=>n=>Bool =
  init = for i j. False
  snd $ withState init \mRef.
    for i:m.
      (from, to) = edges.i
      mRef!from!to := True
```

**Comment by danieldjohnson** [comment]
I think the values `n`, `nodes`, `m`, and `edges` are in scope in the function body but NOT in the type (similar to #285). So it's interpreting the `n` in the return value as a fresh universally quantified type.

More generally I don't think Dex currently supports "lifting" type arguments inside existentially-quantified types (like `Graph`) into non-existential types (like the return value).

**Comment by duvenaud** [comment]
Thanks again, Daniel!  I think you're right.  I had suspected this might be a duplicate of #285, but wasn't sure.

Do you have any intuition about how hard this would be to support?  On the one hand, it might be a bad use of developer time for me to keep coming up with niche, hard-to-support uses of the expressive type system.  On the other hand, the appeal of Dex to me is that, like autograd, I could let my imagination run wild without having to internalize the limitations of a particular implementation.  But I don't have a good intuition for the extent to which the implementation is just missing a few cleanups / refactorings to be much more general, or if it's a Turing tarpit / Russel's paradox mess.  It looks to me to be somewhere in the middle, and that there are a couple of large and somewhat complex features (like subtypes, and allowing functions in type declarations) that will provide a huge amount more generality.

**Comment by danieldjohnson** [comment]
I'm not completely sure, but my intuition is that there would have to be some pretty large-scope redesigns of some of the core typechecking machinery to support this. But on the other hand, I imagine those same redesigns would also open up a bunch of other features as you suggest.

I could be wrong, though, and maybe this specific use case would be fairly doable; I don't have a good mental model of how existential types such as lists are implemented.

**Comment by apaszke** [comment]
This should be resolved by #290!
